modules/playlists/src/main/java/org/opencastproject/playlists/persistence/PlaylistDatabaseServiceImpl.java

```diff
@@ -0,0 +1,237 @@
+        }
+        return playlist.get();
+      });
+    } catch (NotFoundException e) {
+      throw e;
+    } catch (Exception e) {
+      logger.error("Could not retrieve playlist with ID '{}'", playlistId, e);
+      throw new PlaylistDatabaseException(e);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * @see PlaylistDatabaseService#getPlaylists(int, int, SortCriterion)
+   */
+  @Override
+  public List<Playlist> getPlaylists(int limit, int offset, SortCriterion sortCriterion)
+          throws PlaylistDatabaseException {
+
+    try {
+      return db.exec(em -> {
+        CriteriaBuilder criteriaBuilder = em.getCriteriaBuilder();
+        CriteriaQuery<Playlist> criteriaQuery = criteriaBuilder.createQuery(Playlist.class);
+        Root<Playlist> from = criteriaQuery.from(Playlist.class);
+        CriteriaQuery<Playlist> select = criteriaQuery.select(from)
+            .where(criteriaBuilder.isNull(from.get("deletionDate")));
+
+        if (sortCriterion.getOrder().equals(SortCriterion.Order.Ascending)) {
+          criteriaQuery.orderBy(criteriaBuilder.asc(from.get(sortCriterion.getFieldName())));
+        } else if (sortCriterion.getOrder().equals(SortCriterion.Order.Descending)) {
+          criteriaQuery.orderBy(criteriaBuilder.desc(from.get(sortCriterion.getFieldName())));
+        }
+
+        TypedQuery<Playlist> allQuery = em.createQuery(select);
+
+        allQuery.setMaxResults(limit);
+        allQuery.setFirstResult(offset);
+
+        return allQuery.getResultList();
+      });
+    } catch (Exception e) {
+      throw new PlaylistDatabaseException("Error fetching playlists from database", e);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * @see PlaylistDatabaseService#updatePlaylist(Playlist, String)
+   */
+  @Override
+  public Playlist updatePlaylist(Playlist playlist, String orgId) throws PlaylistDatabaseException {
+    try {
+      db.execTx(em -> {
+        Optional<Playlist> fromDb = getPlaylistById(playlist.getId(), orgId).apply(em);
+        playlist.setUpdated(new Date());
+        if (fromDb.isEmpty()) {
+          em.persist(playlist);
+        } else {
+          em.merge(playlist);
+        }
+      });
+      return playlist;
+    } catch (Exception e) {
+      throw new PlaylistDatabaseException("Could not update playlist with ID '" + playlist.getId() + "'", e);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * @see PlaylistDatabaseService#deletePlaylist(Playlist, String)
+   */
+  @Override
+  public Playlist deletePlaylist(Playlist playlist, String orgId) throws PlaylistDatabaseException {
+    try {
+      db.execTx(em -> {
+        Optional<Playlist> fromDb = getPlaylistById(playlist.getId(), orgId).apply(em);
+        if (fromDb.isPresent()) {
+          Date now = new Date();
+          playlist.setUpdated(now);
+          playlist.setDeletionDate(now);
+          em.merge(playlist);
+        }
+      });
+      logger.debug("Playlist with id {} was deleted.", playlist.getId());
```

Comment:
This should move up into the `execTx` block since deleting an already deleted playlist shouldn't do anything, right?